Document DS and event receiver v3 APIs (#402)
dbos-inc/dbos-docs: docs/typescript/reference/plugins.md

---
sidebar_position: 200
title: DBOS Plugin Architecture
---

# DBOS Plugin Architecture

:::tip
Unless you intend to extend the DBOS Transact library, you can ignore this topic.
:::

DBOS Transact for TypeScript currently provides two explicit extension mechanisms:
* [Datasources](#datasources) integrate database clients, ORMs, or other resouces with the DBOS lifecycle and transaction functions.
* [External Event Receivers](#event-receivers) integrate event receivers (pollers, inbound sockets, timers, etc.) with the DBOS lifecycle, and initate workflow functions upon receiving events.

## Datasources

### Method Of Operation

The principle behind DBOS datasources is that DBOS and the datasource work together to execut functions in a checkpointed, transactional context.

> 1. App code calls a transaction function (which may or may not be registered in advance).  Generally, this call is made to the datasource instance, as it provides full type checking for the transaction configuration.

Reviewer: ```suggestion
1. App code calls a transaction function (which does not have to be registered in advance).  Generally, this call is made to the datasource instance, as it provides full type checking for the transaction configuration.
```

6. Control returns to the DBOS step wrapper, which writes the step result to its system database and emits tracing information.
7. The transaction's result is returned to the caller.

### Datasource Interfaces

Datasource integration is based on 3 key interfaces:
 - `DBOSDataSource`: User-facing datasource API.  Part interface and part guideline, this is technically just a suggestion, but following it ensures a consistent developer experience across datasources.
 - `DataSourceTransactionHandler`: This interface is provided by the datasource to DBOS, so that DBOS can control transaction scope.
 - `@dbos-inc/dbos-sdk/datasource`: Exports from this submodule are used by datasources to do register themselves and initiate transactions.

#### `DBOSDataSource`

`DBOSDataSource` is the API for datasources to present to the application developer.  It provides a consistent, type-safe way to register and call transaction functions.
`DBOSDataSource` is templatized on its `Config` type, which describes datasource-specific configuration for a transaction function.  `Config` typically covers transaction isolation, read-write/read-only, and the transaction's name.

>  - `runTransaction` should run its callback, generally with the `runTransaction` function [below](#dbos-incdbos-sdkdatasource)

Reviewer: I don't like how we keep saying "should" and "generally". Maybe something more like: The `DBOSDataSource.runTransaction` method is used to execute arbitrary database code and save the return value in a single transaction. DBOS provides the `runTransaction` helper function described below that can be used when implementing `DBOSDataSource.runTransaction`.

### Associating Information With Classes, Methods, And Parameters
As program initialization is nonlinear, the DBOS library provides infrastructure to help register the classes, functions, and parameters associated with an event receiver.  A cohesive picture of this registration information is then available during `initialize`.

The following `DBOS.` static methods are used to collect registration information during initialization:
```typescript
  associateClassWithInfo(external: AnyConstructor | object | string, cls: AnyConstructor | string): object;

  associateFunctionWithInfo<This, Args extends unknown[], Return>(
    external: AnyConstructor | object | string,
    func: (this: This, ...args: Args) => Promise<Return>,
    target: FunctionName,
  ) : object;

  associateParamWithInfo<This, Args extends unknown[], Return>(
    external: AnyConstructor | object | string,
    func: (this: This, ...args: Args) => Promise<Return>,
    target: FunctionName & {
      param: number | string;
    },
  ) : object | undefined;
```

The parameters to these functions are:

>  - `external`: A key that identifies the event reciever; this can be its constructor, an instance, or a name string.

Reviewer: ```suggestion
 - `external`: A key that identifies the event receiver; this can be its constructor, an instance, or a name string.
```
